[ADD] util/report: add new reporting function

diagnoza opened the pull request:
Add a new function that replaces `util.add_to_migration_reports`.

The new function can automatically parse the provided data as a list of records. It can also limit the number of displayed records to prevent the upgrade report from growing too large.

The structure of the note is now defined within the function itself, standardizing its appearance.

Example usage
===============
More examples will follow in `odoo/upgrade`, replacing the existing use of `util.add_to_migration_reports`.

Example 1
---------------
```
cr.execute(
        """
        SELECT p.id, p.name, p.city, p.comment, p.company_id, com.name FROM res_partner p
          JOIN res_company com
            ON p.company_id = com.id
        """
    )

summary = """
    This is a test of the utility method that is displaying a list of rows
    returned by an SQL query. The number of displayed rows is limited.
    """
row_format = "Partner with id {partner_link} works at company {company_link} in {city}, ({comment})"
columns = ("id", "name", "city", "comment", "company_id", "company_name")
links = {"company_link": ("res.company", "company_id", "company_name"), "partner_link": ("res.partner", "id", "name")}
util.add_report(header="Test of the new utility method", summary=summary, row_format=row_format, columns=columns, links=links, data=cr.fetchall())
```
<img width="1151" height="635" alt="image" src="https://github.com/user-attachments/assets/64542831-f142-484c-becc-ff200297287a" />

Example 2
---------------
```
util.add_report(header="Simple note", summary="A detailed description of the change.")
```
<img width="294" height="75" alt="image" src="https://github.com/user-attachments/assets/aea1c9c6-c7d8-4692-aa82-71ff460e0cdf" />

Example 3
---------------
```
msg = """
This is some text in **bold**.  

This is a line  
that should break.

- Item one
- Item two
    - Sub-item

And some _italic_ text.
"""
util.add_report(header="MD note", summary=msg, report_format="md")
```
<img width="324" height="198" alt="image" src="https://github.com/user-attachments/assets/0c4a9324-cca9-4e4a-834d-ea8ad8d71f63" />


Example 4
---------------
```
msg = """
This is some text in **bold**.

| This is a line
| that should break.

* Item one
* Item two

  * Sub-item

And some *italic* text.
"""
util.add_report(header="RST note", summary=msg, report_format="rst")
```
<img width="411" height="202" alt="image" src="https://github.com/user-attachments/assets/75c92d4e-c1ae-4948-8665-ac4f5c5774ab" />

Remaining details to discuss
===============
1. The introduction of env var to control `limit`.
2. Enforce/assert the use of tuples instead of lists. Currently, it's possible to pass a list for `columns` (instead of a tuple) or a dictionary of lists for `links` (instead of a dictionary of tuples).
3. How to handle depreciation of `add_to_migration_reports`.
4.  Indentation of the `ir.ui.view` record corresponding to the note is sometimes incorrect (previously there was the same issue):
<img width="1527" height="540" alt="image" src="https://github.com/user-attachments/assets/5e9cf10d-2c7f-4898-b842-5bffd1ec231f" />

diagnoza committed:
[ADD] util/report: add new reporting function

Add a new function that replaces util.add_to_migration_reports.

The new function can automatically parse the provided data as a list of records.
It can also limit the number of displayed records to prevent the upgrade report from growing too large.

The structure of the note is now defined within the function itself, standardizing its appearance.

aj-fuentes reviewed (commented):
I think we can do some simplification here. Some suggestions below. Note that none of the suggestions were tested, take it as a guide.

diagnoza commented on src/util/report.py:
```diff
@@ -126,6 +126,95 @@ def add_to_migration_reports(message, category="Other", format="text"):
+            row_text = row_format.format(**row_dict)
+            rows.append("<li>{}</li>".format(row_text))
+
+        rows = "<ul>\n" + "\n".join(rows) + "\n</ul>"
```
Should we explicitly indicate somewhere in the code that this part is added only so that our internal CI can correctly run the test?

aj-fuentes commented on src/util/report.py:
```diff
@@ -126,6 +126,66 @@ def add_to_migration_reports(message, category="Other", format="text"):
+
+    if not data:
+        row_to_html(columns)  # Validate the format is correct, including links
+        return report_with_summary(summary=summary, details="No records to display.", category=category)
```
Maybe this is better:
```suggestion
        return report_with_summary(summary=summary, details="", category=category)
```
then adapt `report_with_summary` to not add any summary/details structure when `details` is empty, and just output summary directly.

aj-fuentes reviewed (commented):
Looks better. Please adapt the documentation to the same format we use elsewhere in this repo. Some suggestions above.

aj-fuentes commented on src/util/report.py:
```diff
@@ -126,6 +126,79 @@ def add_to_migration_reports(message, category="Other", format="text"):
+                                  row_format="Partner with id {partner_link} works at company {company_link} in {city}, ({comment})",
+                                  links={"company_link": ("res.company", "company_id", "company_name"), "partner_link": ("res.partner", "id", "name")},
+                                  category="Accounting"
+                                  )
```
The usual formatting allows for shorter lines. Unless there is some rendering issue?
```suggestion
            util.report_with_list(
                summary="The following records were altered.",
                data=cr.fetchall(),
                columns=("id", "name", "city", "comment", "company_id", "company_name")
                row_format="Partner with id {partner_link} works at company {company_link} in {city}, ({comment})",
                links={"company_link": ("res.company", "company_id", "company_name"), "partner_link": ("res.partner", "id", "name")},
                category="Accounting"
            )
```

KangOl commented on src/util/report.py:
```diff
@@ -126,6 +126,79 @@ def add_to_migration_reports(message, category="Other", format="text"):
+        return "<li>{}</li>".format(row_format.format(**row_dict))
+
+    if not data:
+        row_to_html(columns)  # Validate the format is correct, including links
```
It will fail. `column` doesn't have the expected type.

(a.k.a tests are missing)

diagnoza commented on src/base/tests/test_util.py:
```diff
@@ -2677,3 +2677,63 @@ def test_assert_updated_combo(self):
+                None,
+                None,
+                100,
+                "Other",
```
I had to pass the default arguments as regular ones, because of how the tested method is called in `test_report_with_list`. Simply leaving them out would obviously result in
```
Traceback (most recent call last):
  File "/home/odoo/src/upgrade-util/src/testing.py", line 174, in wrapped
    return func(self, *args)
TypeError: TestReportUtils.test_report_with_list() missing 2 required positional arguments: 'category' and 'expected'
```

Passing them as `None` also wouldn't work, because since Python differentiates between "no argument passed" and `None`, they would actually be set to `None` instead of taking the default value. And if we pass `category=None`, then `add_to_migration_reports` will break on running `len(None)`.

diagnoza replied:
No longer an issue since we don't use `paramterize`.

KangOl commented on src/base/tests/test_util.py:
```diff
@@ -2682,3 +2682,63 @@ def test_assert_updated_combo(self):
+
+
+class TestReportUtils(UnitTestCase):
+    @parametrize(
```
In this case, using `paramatrize` make the test less readable.
Better directly write 3 test functions.

diagnoza committed:
[IMP] util/fields: use the new report util

Use the new utility function designed specifically for
adding a report with a list of records.

diagnoza commented on src/util/report.py:
```diff
@@ -127,6 +127,90 @@ def add_to_migration_reports(message, category="Other", format="text"):
+
+    disclaimer = "The total number of affected records is {}. ".format(total) if total else ""
+    total = len(data) if total is None else total
+    limit = min(limit, total) if limit != -1 else total
```
As per DM with Alvaro:

1. We shouldn't assume devs don't know the language they're using.
2. `None` is actually more descriptive than `-1` for "no limit".

diagnoza commented:
@aj-fuentes @KangOl it's ready for another look.

diagnoza commented:
upgradeci retry

aj-fuentes commented on src/util/report.py:
```diff
@@ -127,6 +127,90 @@ def add_to_migration_reports(message, category="Other", format="text"):
+                    link: get_anchor_link_to_record(rec_model, row_dict[id_col], row_dict[name_col])
+                    for link, (rec_model, id_col, name_col) in links.items()
+                }
+            )
```
Once we generate the link items (which are already [escaped](https://github.com/odoo/upgrade-util/blob/0e5bd206f460681c7de3203597b77ff82b2181c0/src/util/report.py#L339)) I think we need to escape the remaining values in the dict.
For example imagine that `row_format` is `"{some_field}"` and in the dict we get `{"some_field": "<script>window.alert('escaped')</script>"}`
```suggestion
            )
        row_dict = {col: html_escape(val) for col, val in row_dict.items()}
```
cc: @KangOl

aj-fuentes commented on src/util/fields.py:
```diff
@@ -1598,11 +1598,26 @@ def _update_field_usage_multi(cr, models, old, new, domain_adapter=None, skip_in
+            "{model_text}: the field <kbd>{old}</kbd> has been renamed to <kbd>{new}</kbd>. "
+            "The following server actions may need an update. "
+            "If a server action is a standard one and you haven't made any modifications, you may ignore it."
+        ).format(model_text=model_text, old=old, new=new)
```
We can use triple-quote strings here, in the end we are formatting HTML.
```suggestion
        summary = """
            {model_text}: the field <kbd>{old}</kbd> has been renamed to <kbd>{new}</kbd>.
            The following server actions may need an update.
            If a server action is a standard one and you haven't made any modifications, you may ignore it.
        """.format(model_text=model_text, old=old, new=new)
```

aj-fuentes commented on src/util/fields.py:
```diff
@@ -1621,28 +1636,24 @@ def _update_field_usage_multi(cr, models, old, new, domain_adapter=None, skip_in
+                "{model_text}: the field <kbd>{old}</kbd> has been renamed to <kbd>{new}</kbd>. "
+                "The following compute methods of other fields may need an update. "
+                "If a field is a standard one and you haven't made any modifications, you may ignore it."
+            ).format(model_text=model_text, old=old, new=new)
```
```suggestion
            summary = """
                {model_text}: the field <kbd>{old}</kbd> has been renamed to <kbd>{new}</kbd>.
                The following compute methods of other fields may need an update.
                If a field is a standard one and you haven't made any modifications, you may ignore it.
            """.format(model_text=model_text, old=old, new=new)
```

aj-fuentes commented on src/util/report.py:
```diff
@@ -127,6 +127,91 @@ def add_to_migration_reports(message, category="Other", format="text"):
+
+    def row_to_html(row):
+        raw = dict(zip(columns, row))
+        row_dict = {col: html_escape(val) for col, val in raw.items()}
```
`val` could be non-str.
```suggestion
        row_dict = {col: html_escape(str(val)) for col, val in raw.items()}
```

KangOl commented on src/util/report.py:
```diff
@@ -107,7 +107,7 @@ def html_escape(text):
 
 
-def add_to_migration_reports(message, category="Other", format="text"):
+def report(message, category="Other", format="text"):
```
That's not retro-compatible as it now shadow the `report` package.
The following snippet will not work anymore.
```python
from odoo.upgrade import util

util.report.html_escape("whatever")
```

```suggestion
def report_message(message, category="Other", format="text"):
```

KangOl commented:
Then, commit message should be adapted to the new function name.

KangOl commented on src/util/report.py:
```diff
@@ -107,7 +107,7 @@ def html_escape(text):
 
 
-def add_to_migration_reports(message, category="Other", format="text"):
+def report_message(message, category="Other", format="text"):
```
Other (stupid) proposals

```suggestion
def tell_customer(message, category="Other", format="text"):
```
```suggestion
def inform_user(message, category="Other", format="text"):
```
```suggestion
def yell(message, category="Other", format="text"):
```
```suggestion
def beware(message, category="Other", format="text"):
```
```suggestion
def upgrade_awareness(message, category="Other", format="text"):
```

KangOl reviewed (approved):
@robodoo rebase-ff r+

![](https://pixelfed.social/storage/m/_v2/435438564597290694/4c90d8e69-c5a1c2/WHw8ibX0YnMr/Jrz79Mdsz53YFtOsH1Bh5BqCrARxhdFwvcwxpuQf.jpg)[^source]

[^source]: https://pixelfed.social/p/headofcatdept/980370589055479753

robodoo commented:
Merge method set to rebase and fast-forward.
